cmd/internal/start_components.go

```diff
@@ -0,0 +1,153 @@
+package internal
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/lightstep/opentelemetry-prometheus-sidecar/config"
+	"github.com/lightstep/opentelemetry-prometheus-sidecar/metadata"
+	"github.com/lightstep/opentelemetry-prometheus-sidecar/otlp"
+	"github.com/lightstep/opentelemetry-prometheus-sidecar/prometheus"
+	"github.com/lightstep/opentelemetry-prometheus-sidecar/retrieval"
+	"github.com/lightstep/opentelemetry-prometheus-sidecar/tail"
+	"github.com/oklog/run"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/tsdb/wal"
+)
+
+type SidecarConfig struct {
+	ClientFactory otlp.StorageClientFactory
+	Monitor       *prometheus.Monitor
+	Logger        log.Logger
+	InstanceId    string
+	Filters       [][]*labels.Matcher
+	MetricRenames map[string]string
+	MetadataCache *metadata.Cache
+
+	config.MainConfig
+}
+
+// createPrimaryDestinationResourceLabels returns the OTLP resources
+// to use for the primary destination.
+func createPrimaryDestinationResourceLabels(svcInstanceId string, extraLabels map[string]string) labels.Labels {
+	// Note: there is minor benefit in including an external label
+	// to indicate the process ID here.  See
+	// https://github.com/lightstep/opentelemetry-prometheus-sidecar/issues/44
+	// Until resources are serialized once per request, leave this
+	// commented out (and a test in e2e_test.go):
+	// extraLabels[externalLabelPrefix+string(semconv.ServiceInstanceIDKey)]
+	// = svcInstanceId
+	return labels.FromMap(extraLabels)
+}
+
+func NewTailer(ctx context.Context, scfg SidecarConfig) (*tail.Tailer, error) {
+	return tail.Tail(
+		ctx,
+		log.With(scfg.Logger, "component", "wal_reader"),
+		scfg.Prometheus.WAL,
+		scfg.Monitor,
+	)
+}
+
+func StartComponents(ctx context.Context, scfg SidecarConfig, tailer tail.WalTailer, startOffset int) error {
+	var err error
+	attempts := 0
+	currentSegment := 0
+	for {
+		currentSegment, err = runComponents(ctx, scfg, tailer, startOffset)
+		if err != nil && attempts < config.DefaultMaxRetrySkipSegments && strings.Contains(err.Error(), tail.ErrSkipSegment.Error()) {
+			_ = tailer.Close()
+			_ = retrieval.SaveProgressFile(scfg.Prometheus.WAL, startOffset)
+			tailer, err = NewTailer(ctx, scfg)
+			if err != nil {
+				level.Error(scfg.Logger).Log("msg", "tailing WAL failed", "err", err)
+				break
+			}
+			attempts += 1
+			tailer.SetCurrentSegment(currentSegment)
```

Comment:
This `SetCurrentSegment()` is suspicious to me, because the `NewTailer()` will find the `wal.LastCheckpoint()` and set next segment automatically. Can we remove it?

My guess is that this is working out because `currentSegment` happens to be the value we expect.

To explain this, the WAL truncation event happens when segment `K` is replaced by checkpoint `K` so that either way, the correct next segment is always K+1. We are seeing a truncation in segment `K` and re-opening with checkpoint `K`.

Will you add a log statement in `tail.Tail()` to indicate which checkpoint is being opened?

Reply:
I've definitely seen the situation where K is truncated and it isn't because of a checkpoint occurring, which is why I left this code in. Otherwise a new tailer will always hit that same truncated segment

Reply:
Great. Thanks for the explanation. Will you add a comment summarizing this thread in the code (since it's really not obvious how checkpoints are read)? Thanks!